File: src/main/java/com/ramya/ecomerce/service/ProductService.java
package com.ramya.ecomerce.service;

import com.ramya.ecomerce.entity.Product;
import com.ramya.ecomerce.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    public Product getProductById(Long id) {
        return productRepository.findById(id).orElse(null);
    }

    public List<Product> searchProducts(String search) {
        if (search == null || search.isEmpty()) {
            return getAllProducts();
        }
        return productRepository.searchProducts(search);
    }

    public List<Product> filterProducts(Double minPrice, Double maxPrice, Double minRating, Double maxRating) {
        return productRepository.filterProducts(minPrice, maxPrice, minRating, maxRating);
    }

    public List<Product> sortProducts(List<Product> products, String sortBy, String order) {
        Comparator<Product> comparator = null;

        if ("price".equalsIgnoreCase(sortBy)) {
            comparator = Comparator.comparingDouble(Product::getPrice);
        } else if ("rating".equalsIgnoreCase(sortBy)) {
            comparator = Comparator.comparingDouble(Product::getRating);
        } else if ("name".equalsIgnoreCase(sortBy)) {
            comparator = Comparator.comparing(Product::getName);
        }

        if (comparator != null) {
            if ("desc".equalsIgnoreCase(order)) {
                comparator = comparator.reversed();
            }
            return products.stream().sorted(comparator).collect(Collectors.toList());
        }

        return products;
    }
}

File: frontend/src/services/productService.js
import API from './api';

export const productService = {
  getAllProducts: async () => {
    const response = await API.get('/products');
    return response.data;
  },

  getProductById: async (id) => {
    const response = await API.get(`/products/${id}`);
    return response.data;
  },

  searchProducts: async (search) => {
    const response = await API.get('/products', { params: { search } });
    return response.data;
  },

  filterProducts: async (minPrice, maxPrice, minRating, maxRating) => {
    const response = await API.get('/products', {
      params: { minPrice, maxPrice, minRating, maxRating }
    });
    return response.data;
  },

  sortProducts: async (sortBy, order) => {
    const response = await API.get('/products', { params: { sortBy, order } });
    return response.data;
  }
};

